From 32a9bc9df09ca9741b44a356cd6ca4e4c744f3d5 Mon Sep 17 00:00:00 2001
From: Weiwei Wang <weiwei.wang@windriver.com>
Date: Tue, 11 Jan 2011 19:58:02 +0800
Subject: [PATCH 2/4] ppc preempt-rt: delete local_irq_enable in do_signal_pending

irq is already enabled by entry_32.S and entry_64.S before
do_signal_pending is called, so when CONFIG_PROVE_LOCKING is
enabled, the following call trace will be produced:
Badness at /buildarea3/wwang0/5build/p4080-preempt/build/linux/kernel/lockdep.c:2352
NIP: c007eed8 LR: c007eec0 CTR: c000e12c
REGS: eb647da0 TRAP: 0700   Not tainted  (2.6.34.6-rt-WR4.0.0.0_preempt_rt)
MSR: 00029002 <EE,ME,CE>  CR: 28222422  XER: 20000000
TASK = ea5bd540[1131] 'cpu_load_multi.' THREAD: eb646000 CPU: 4
GPR00: 00000000 eb647e50 ea5bd540 00000001 00000002 28222424 00000004 0ff13410
GPR08: 0002d002 c0760000 00000001 c0760000 0002d002 100a7a08 00000060 1009fd40
GPR16: 100a507c 00000000 1008a86c bfd4a540 00000034 00000004 1009feb8 0000060a
GPR24: 0000060c 00000003 bfd4a540 00000001 00000080 eb647f50 c0009194 ea5bd540
NIP [c007eed8] trace_hardirqs_on_caller+0x1c4/0x1cc
LR [c007eec0] trace_hardirqs_on_caller+0x1ac/0x1cc
Call Trace:
[eb647e50] [c00817d0] lock_acquire+0x88/0x124 (unreliable)
[eb647e70] [c0009194] T.460+0x18/0x270
[eb647f30] [c000943c] do_signal+0x50/0x58
[eb647f40] [c001203c] do_user_signal+0x74/0xc4
Instruction dump:
801c0040 70090004 4182fe90 484a5b1d 4bfffe88 4821e8a5 2f830000 41befefc
3d20c076 800993c8 2f800000 40befeec <0fe00000> 4bfffee4 7c6802a6 4bfffe30

Here delete redundant local_irq_enable to avoid it.

Signed-off-by: Weiwei Wang <weiwei.wang@windriver.com>
---
 arch/powerpc/kernel/signal.c |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/arch/powerpc/kernel/signal.c b/arch/powerpc/kernel/signal.c
index de4960c..bdb52b7 100644
--- a/arch/powerpc/kernel/signal.c
+++ b/arch/powerpc/kernel/signal.c
@@ -122,9 +122,8 @@ static int do_signal_pending(sigset_t *oldset, struct pt_regs *regs)
 
 #ifdef CONFIG_PREEMPT_RT
 	/*
-	 * Fully-preemptible kernel does not need interrupts disabled:
+	 * Add a preempt sched port
 	 */
-	local_irq_enable();
 	preempt_check_resched();
 #endif
 
-- 
1.6.5.2

